### PR TITLE
Fix memory leak in EditorConfigDocumentOptionsProvider

### PR DESCRIPTION
**Pull request by sharwell**
This pull request implements the memory leak fix from #26386.

### Customer scenario

A memory leak occurs if a solution is closed while documents are open.

### Bugs this fixes

This is an interim solution for #26377.

### Workarounds, if any

Either of the following:

* Close all documents before closing a solution
* Restart the IDE when opening a different solution or reopening the same solution

### Risk

Low. The code only impacts cases where a project or solution is getting closed, and takes the same actions that occur if we received the expected events for closing documents.

### Performance impact

Minor performance improvement in the impacted scenario.

### Is this a regression from a previous update?

No.

### Root cause analysis

The leak is minor (a few small object graphs) so it didn't show up in profiling reports.

### How was the bug found?

Found by @heejaechang the investigation for #15003.

### Test documentation updated?

N/A

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs [hunk 1]
```diff
@@ -63,6 +65,69 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
...
+                            _openDocumentContexts.Remove(documentId);
+                        }
+                    }
+                }
+            }
+        }
+
+        private void ReleaseContext_NoLock(Task<ICodingConventionContext> contextTask)
+        {
+            Debug.Assert(Monitor.IsEntered(_gate));
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: Slick. Not sure why we don't do this in more places.

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs [hunk 2]
```diff
@@ -63,6 +65,69 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
             }
         }
 
+        private void Workspace_WorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            switch (e.Kind)
+            {
+            case WorkspaceChangeKind.SolutionRemoved:
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: Don't we indent cases? Not sure if bug or unstated .editorconfig rule.

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs [hunk 3]
```diff
@@ -63,6 +65,69 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
...
+                // If any items were released due to the Clear operation, we need to remove them from the map.
+                if (itemReleased)
+                {
+                    if (projectId is null)
+                    {
+                        _openDocumentContexts.Clear();
+                    }
+                    else
+                    {
+                        var itemsToRemove = _openDocumentContexts.Keys.Where(key => key.ProjectId == projectId).ToList();
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: If we're still creating a list here, why not just add the keys to release in a list in the first loop and avoid having the duplicate logic? I get we can avoid the list in the close solution case, but the number of open files isn't usually going to be big enough for that to matter.

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: Unstated rule. Will be fixed in 32e150c9ae85d9ad3e3233ea374e3ae0cab26d3b (different PR).

Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: ➡️ Amended the initial commit to fix the formatting.

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs (hunk 3 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: ➡️ Replaced with simpler logic in 0473ba4f0089e72a052147f7efeeb6818c03c5c5.

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs]: 👍 Now we know 😄

**Pull request merged**